PR: https://github.com/odoo/odoo/pull/223606

From: 18b6ed1cf7649cca26cebfa2733daee6edc01111
From: Lina (liew)
Date: 2025-08-21 13:38:51

Breaking data model changes scores: del:2 + add:0, change matches:
-    paymob_access_token = fields.Char(groups='base.group_system')
-    paymob_access_token_expiry = fields.Datetime(default='1970-01-01', groups='base.group_system')

Total Changes: 29

[FIX] payment_paymob: fetch access token always

If an access token is fetched, then it won't be invalidated
when new credentials are given and it might result in faulty behavior.

Since the expiry of the access token is about 60 mins, then there is
no value to storing the access token on the payment provider and can be fetched when needed.

closes odoo/odoo#223606

Related: odoo/upgrade#8258
Signed-off-by: Lina Eweis (liew) <liew@odoo.com>

================================= pseudo patch: =================================

--- a/addons/payment_paymob/models/payment_provider.py
+++ b/addons/payment_paymob/models/payment_provider.py
@@ -2,7 +2,7 @@
 
 import json
 
-from datetime import datetime, timedelta
+from datetime import datetime
 
 from odoo import _, api, fields, models
 from odoo.exceptions import ValidationError
@@ -36,8 +36,6 @@ class PaymentProvider(models.Model):
     )
     paymob_hmac_key = fields.Char(string="Paymob HMAC Key", required_if_provider='paymob')
     paymob_api_key = fields.Char(string="Paymob API Key", required_if_provider='paymob')
-    paymob_access_token = fields.Char(groups='base.group_system')
-    paymob_access_token_expiry = fields.Datetime(default='1970-01-01', groups='base.group_system')
 
     # === CONSTRAINT METHODS === #
 
@@ -232,21 +230,16 @@ class PaymentProvider(models.Model):
         :rtype: str
         :raise ValidationError: If the access token can not be fetched.
         """
-        if not self.paymob_access_token or fields.Datetime.now() > self.paymob_access_token_expiry:
-            response_content = self._send_api_request(
-                'POST',
-                '/api/auth/tokens',
-                json={'api_key': self.paymob_api_key},
-                is_refresh_token_request=True,
-            )
-            access_token = response_content['token']
-            if not access_token:
-                raise ValidationError(_("Could not generate a new access token."))
-            self.write({
-                'paymob_access_token': access_token,
-                'paymob_access_token_expiry': fields.Datetime.now() + timedelta(minutes=55),
-            })
-        return self.paymob_access_token
+        response_content = self._send_api_request(
+            'POST',
+            '/api/auth/tokens',
+            json={'api_key': self.paymob_api_key},
+            is_refresh_token_request=True,
+        )
+        access_token = response_content['token']
+        if not access_token:
+            raise ValidationError(_("Could not generate a new access token."))
+        return access_token
 
     def _parse_response_error(self, response):
         """Override of `payment` to parse the error message."""
